### PR TITLE
Rework fetch_bytes and _fetch_pypi_json patches to make testing easier

### PR DESCRIPTION
**Pull request by hoodmane**
Various improvements to test_micropip. The main feature is a new fixture
'mock_fetch' with an add_pkg method that takes a package name, a map
from version to requirements, and a choice of platform. This should
hopefully make writing more tests a lot easier (which is good because
we could use more micropip test coverage but we are limited by the
difficulty of writing good tests).

This also adds a fixture to create distinct dummy package names and
enables @pytest.mark.asyncio to handle the async calls rather than
using `asyncio.get_event_loop.run_until_complete`.

### CODE REVIEW

**Review by ryanking13 — COMMENTED** [review]
Thanks for this work @hoodmane!

File: packages/micropip/src/micropip/_micropip.py [hunk 1]
```diff
@@ -36,7 +36,7 @@ async def _get_pypi_json(pkgname: str, fetch_kwargs: dict[str, str]) -> Any:
     url = f"https://pypi.org/pypi/{pkgname}/json"
     try:
         metadata = await fetch_string(url, fetch_kwargs)
-    except Exception as e:
+    except OSError as e:
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply packages/micropip/src/micropip/_micropip.py:39]: We raise `OSError` here:
https://github.com/pyodide/pyodide/blob/main/src/py/pyodide/http.py#L107
Thankfully this has test coverage: I tried to put `JSException` first and found out it was raising `OSError`.

[the remark this answers is not in the document]
Comment (ryanking13) [reply packages/micropip/src/micropip/_micropip.py:39]: Hmm maybe the 4xx error is not catched there? I get this error in our latest console.

```py
>>> import micropip
>>> await micropip.install("raise404error")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/lib/python3.10/asyncio/futures.py", line 284, in __await__
    yield self  # This tells Task to wait for completion.
  File "/lib/python3.10/asyncio/tasks.py", line 304, in __wakeup
    future.result()
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/lib/python3.10/asyncio/tasks.py", line 234, in __step
    result = coro.throw(exc)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 183, in install
    transaction = await self.gather_requirements(requirements, ctx, keep_going)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 173, in gather_requirements
    await gather(*requirement_promises)
  File "/lib/python3.10/asyncio/futures.py", line 284, in __await__
    yield self  # This tells Task to wait for completion.
  File "/lib/python3.10/asyncio/tasks.py", line 304, in __wakeup
    future.result()
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/lib/python3.10/asyncio/tasks.py", line 234, in __step
    result = coro.throw(exc)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 280, in add_requirement
    metadata = await _get_pypi_json(req.name)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 86, in _get_pypi_json
    return json.loads(await fetch_string(url))
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 58, in fetch_string
    return await (await pyfetch(url, **kwargs)).string()
  File "/lib/python3.10/site-packages/pyodide/http.py", line 231, in pyfetch
    url, await _jsfetch(url, to_js(kwargs, dict_converter=Object.fromEntries))
  File "/lib/python3.10/asyncio/futures.py", line 284, in __await__
    yield self  # This tells Task to wait for completion.
  File "/lib/python3.10/asyncio/tasks.py", line 304, in __wakeup
    future.result()
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception
JsException: TypeError: Failed to fetch
```

[the remark this answers is not in the document]
Comment (ryanking13) [reply packages/micropip/src/micropip/_micropip.py:39]: It looks like fetch raises an exception when it gets 4xx, 5xx from the response.

[the remark this answers is not in the document]
Comment (hoodmane) [reply packages/micropip/src/micropip/_micropip.py:39]: Okay, I think we should update `pyfetch` then? I think the intended behavior was to raise `OSError`.

[the remark this answers is not in the document]
Comment (ryanking13) [reply packages/micropip/src/micropip/_micropip.py:39]: > Okay, I think we should update `pyfetch` then? I think the intended behavior was to raise `OSError`.

Yeah, I'm +1 for it.

[the remark this answers is not in the document]
Comment (hoodmane) [reply packages/micropip/src/micropip/_micropip.py:39]: The issue is that

**Pull request merged**